source/Nevermore/IQueryBuilder.cs

```diff
@@ -1,40 +1,85 @@
 using System;
 using System.Collections.Generic;
-using Nevermore.Joins;
+using Nevermore.AST;
 
 namespace Nevermore
 {
-    public interface IQueryBuilder<TRecord> where TRecord : class
+    public interface IOrderedQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+    }
+
+    public interface ITableSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
     {
-        IQueryGenerator QueryGenerator { get; }
+        ITableSourceQueryBuilder<TRecord> View(string viewName);
+        ITableSourceQueryBuilder<TRecord> Table(string tableName);
+        ITableSourceQueryBuilder<TRecord> Alias(string tableAlias); 
+        IQueryBuilder<TRecord> Hint(string tableHint);
+        IAliasedSelectSource AsAliasedSource();
+    }
 
+    public interface ISubquerySourceBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        ISubquerySource AsSource();
+        ISubquerySourceBuilder<TRecord> Alias(string subqueryAlias);
+    }
+
+    public interface IJoinSourceQueryBuilder<TRecord> : IQueryBuilder<TRecord> where TRecord : class
+    {
+        IJoinSourceQueryBuilder<TRecord> On(string leftField, JoinOperand operand, string rightField);
+    }
+
+    public interface IQueryBuilder<TRecord> where TRecord : class
+    {
         IQueryBuilder<TRecord> Where(string whereClause);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object value);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, object startValue, object endValue);
-        IQueryBuilder<TRecord> Where(string fieldName, SqlOperand operand, IEnumerable<object> values);
-        
+        IQueryBuilder<TRecord> WhereParameterised(string fieldName, UnarySqlOperand operand, Parameter parameter);
```

Comment:
Why change the name of this method? Isn't it going to make it harder to update all the code that uses it?

I'm also not sure what you mean by `Parameterised` as `Where(string)` can also (and usually is) parameterised

Reply:
One of the aims is to be able to to use this query builder for stored procs / functions. Which means `Parameter` is a little bit overloaded. It means
1. The `@` parameter in the sql query string
2. The actual value associated with the `@` parameter that gets used at execution time

For 1. we need a way of parameterizing the sql query without providing a value (existing code makes you always provide a value). The new method `WhereParameterised` allows you to specify a where clause with a parameter but __without__ a value.

Any existing usages will switch to using extension methods, so don't need to change at all, and these extension method are implemented by:
1. Calling `WhereParameterised` (which allows us to share the same code for parameterising where clause in stored procs and normal queries)
2. Adding the actual parameter value to the `CommandParameterValues` collection

I imagine that it is mostly Stored Procs and Functions that would use `WhereParameterised` directly instead of the `Where` extension methods, so one reason that I named this method `WhereParameterised` was to make it distinct from `Where` and discourage people from using it in normal queries.

Reply:
`Where(string fieldName, SqlOperand operand, object value)` and `Where(string fieldName, SqlOperand operand, IEnumerable<object> values)` should quite commonly used since we introduced them. We've been encouraging people to steer away from `Where(string)` as the former can be strongly typed and automatically hook up the right parameters, eg `Where(nameof(Project.Name), SqlOperand.Equals, "foo")`

Reply:
Yep, and I haven't made any changes to that. People can still create Where expressions with `Where(string)` if they want to shoot themselves in the foot, or they can use extension methods like `Where(string fieldName, SqlOperand operand, object value)`